subsys/bluetooth/host/l2cap.c

```diff
@@ -1351,12 +1347,62 @@ static void l2cap_chan_update_credits(struct bt_l2cap_le_chan *chan,
 
 	bt_l2cap_send(chan->chan.conn, BT_L2CAP_CID_LE_SIG, buf);
 
-done:
 	BT_DBG("chan %p credits %u", chan, k_sem_count_get(&chan->rx.credits));
 }
 
-static struct net_buf *l2cap_alloc_frag(struct bt_l2cap_le_chan *chan)
+static void l2cap_chan_update_credits(struct bt_l2cap_le_chan *chan,
+				      struct net_buf *buf)
+{
+	s16_t credits;
+
+	/* Restore enough credits to complete the sdu */
+	if (chan->_sdu) {
+		credits = ((chan->_sdu_len - net_buf_frags_len(chan->_sdu)) +
+			  (chan->rx.mps - 1)) / chan->rx.mps;
+	} else {
+		credits = chan->rx.init_credits;
+	}
+
+	credits -= k_sem_count_get(&chan->rx.credits);
+	if (credits <= 0) {
+		return;
+	}
+
+	l2cap_chan_send_credits(chan, buf, credits);
+}
+
+int bt_l2cap_chan_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
+{
+	struct bt_l2cap_le_chan *ch = BT_L2CAP_LE_CHAN(chan);
+	struct bt_conn *conn = chan->conn;
+	u16_t credits;
+
+	__ASSERT_NO_MSG(chan);
+	__ASSERT_NO_MSG(buf);
+
+	if (!conn) {
+		return -ENOTCONN;
+	}
+
+	if (conn->type != BT_CONN_TYPE_LE) {
+		return -ENOTSUP;
+	}
+
+	BT_DBG("chan %p buf %p", chan, buf);
+
+	/* Restore credits used by packet */
+	credits = (net_buf_frags_len(buf) + (ch->rx.mps + 1)) / ch->rx.mps;
```

Comment:
Is it +1 because you simplified this?
net_buf_frags_len(buf) + 2 + (ch->rx.mps - 1)) / ch->rx_mps
If so I think it is better to write it like that.

Reply:
Yep it is counting the SDU length I will add that.